cpp/mrc/src/tests/test_service.cpp

```diff
@@ -0,0 +1,405 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2021-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "tests/common.hpp"
+
+#include "internal/service.hpp"
+
+#include "mrc/exceptions/runtime_error.hpp"
+
+#include <glog/logging.h>
+#include <gtest/gtest.h>
+
+#include <atomic>
+
+namespace mrc {
+
+class SimpleService : public Service
+{
+  public:
+    SimpleService(bool do_call_in_destructor = true) :
+      Service("SimpleService"),
+      m_do_call_in_destructor(do_call_in_destructor)
+    {}
+
+    ~SimpleService() override
+    {
+        if (m_do_call_in_destructor)
+        {
+            Service::call_in_destructor();
+        }
+    }
```

Comment:
Question: If `m_do_call_in_destructor` and `call_in_destructor` are both declared in `Service`, why not have `Service`'s destructor container this logic?

Reply:
`call_in_destructor` can call the virtual methods for stopping and killing. So you cant execute virtual functions in a base class destructor.